src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketsTelemetry.cs

```diff
@@ -0,0 +1,57 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics.Tracing;
+
+namespace System.Net.Sockets
+{
+    [EventSource(Name = "System.Net.Sockets")]
+    internal sealed class SocketsTelemetry : EventSource
+    {
+        public static readonly SocketsTelemetry Log = new SocketsTelemetry();
+
+        [NonEvent]
+        public static bool IsEnabled(EventLevel level)
+        {
+            return Log.IsEnabled(level, EventKeywords.All);
+        }
+
+        [NonEvent]
+        public void ConnectStart(Internals.SocketAddress address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [NonEvent]
+        public void ConnectStart(EndPoint address)
+        {
+            ConnectStart(address.ToString());
+        }
+
+        [Event(1, Level = EventLevel.Informational)]
+        public void ConnectStart(string? address)
+        {
+            WriteEvent(eventId: 1, address ?? "");
+        }
+
+        [Event(2, Level = EventLevel.Informational)]
+        public void ConnectStop()
+        {
+            WriteEvent(eventId: 2);
+        }
+
+        [Event(3, Level = EventLevel.Error)]
+        public void ConnectFailed()
+        {
+            WriteEvent(eventId: 3);
+            ConnectStop();
```

Comment:
This is a bit non-standard.  Generally if you want to trigger two events, you have a `[NonEvent]` method that calls both of them.  Generally speaking, `[Event]` methods should only log the event that they represent.

Reply:
Can it lead to invalid log entries? It was done per above request to avoid unnecessary call to ConnectStop after ConnectFailed. I could have introduced a new private method like ConnectStopInternal that will be called from both of ConnectStop and ConnectFailed, but it looked redudant.

Reply:
I can't think of any reason why this would cause corruption or invalid log entries, but it is certainly non-standard.  If I understood the comment above, the goal was to not have two calls to it, one in each branch, and rather just call it afterwards unconditionally (for cleanliness).  I would recommend doing something like that rather than this to avoid chaining these events together.  The more you follow the standard `EventSource` rules, the better the chance things don't break in the future.  I can't point to anything that I think would break now though.

Reply:
Extracted connection stop logging into a new ConnectionStopInternal method to be on the safe side.